src/System.IO.FileSystem/src/System/IO/UnixFileSystemObject.cs

```diff
@@ -265,20 +265,16 @@ private void EnsureStatInitialized()
                     int errno = _fileinfoInitialized;
                     _fileinfoInitialized = -1;
 
-                    bool failedBecauseOfDirectory = _isDirectory;
-                    if (!failedBecauseOfDirectory && errno == Interop.Errors.ENOENT)
-                    {
-                        // Windows distinguishes between whether the directory or the file isn't found,
-                        // and throws a different exception in these cases.  We attempt to approximate that
-                        // here; there is a race condition here, where something could change between
-                        // when the error occurs and our checks, but it's the best we can do, and the
-                        // worst case in such a race condition (which could occur if the file system is
-                        // being manipulated concurrently with these checks) is that we throw a
-                        // FileNotFoundException instead of DirectoryNotFoundexception.
-                        failedBecauseOfDirectory = !Directory.Exists(Path.GetDirectoryName(_fullPath));
-                    }
-
-                    throw Interop.GetExceptionForIoErrno(errno, _fullPath, failedBecauseOfDirectory);
+                    // Windows distinguishes between whether the directory or the file isn't found,
+                    // and throws a different exception in these cases.  We attempt to approximate that
+                    // here; there is a race condition here, where something could change between
+                    // when the error occurs and our checks, but it's the best we can do, and the
+                    // worst case in such a race condition (which could occur if the file system is
+                    // being manipulated concurrently with these checks) is that we throw a
+                    // FileNotFoundException instead of DirectoryNotFoundexception.
+                    string parent = Path.GetDirectoryName(PathHelpers.TrimEndingDirectorySeparator(_fullPath));
+                    bool directoryError = (errno == Interop.Errors.ENOENT && !Directory.Exists(parent)); // The destFile's path is invalid
```

Comment:
The main thing I wanted to change was the defaulting of failedbecauseofDirectory to _isDirectory. The Windows behavior will throw a FileNotFoundException if the file name is invalid, even if the object represented by the FileSystemObject is actually a Directory. DirectoryNotFoundExceptions are thrown when the path to the object is invalid. After changing that, I refactored it to this way but you're correct that the string allocations are unnecessary.